### PR TITLE
Joneskm/ibc

### PR DESCRIPTION
**Pull request by joneskm**
IBC resurrection

### CODE REVIEW

File: database/src/rocks.rs [hunk 1]
```diff
@@ -25,7 +25,7 @@ impl Database for RocksDB {
     fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
         self.db
             .get(key)
-            .unwrap_or_else(|e| panic!("unrecoverable database error {}", e))
+            .unwrap_or_else(|e| panic!("unrecoverable database error {}", e)) //TODO: this is probably not the right thing to do when handling a abci or REST query. Perhaps we should have a must_get method?
```
Comment (NYBACHOK) [inline database/src/rocks.rs:28]: How it would differ from this `get`?

File: Makefile [hunk 2]
```diff
@@ -18,4 +18,13 @@ init:
 tendermint-start:
 	tendermint start --home ~/.gaia-rs
 
-.PHONY: run run-debug test install init tendermint-start
\ No newline at end of file
+init-second:
```
Comment (NYBACHOK) [inline Makefile:21]: Second?

File: docs/tutorials/ibc.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+```txt
+Got query request to: store/ibc/key
+```
+
+### Opening a connection
+
+This requires keys:
+
+```shell
+hermes keys restore test-chain --mnemonic "race draft rival universe maid cheese steel logic crowd fork comic easy truth drift tomorrow eye buddy head time cash swing swift midnight borrow"
```
Comment (NYBACHOK) [inline docs/tutorials/ibc.md:76]: Is it expected to know what is `hermes`? I suppose that in future there will be more instructions on, but if we depend on specific version (as we do with tendermint) than you need to leave link to this.

File: store/src/types/prefix/mutable.rs [hunk 4]
```diff
@@ -7,7 +7,7 @@ use crate::{
 
 /// Wraps an mutable KVStore with a prefix
 pub struct MutablePrefixStore<'a, DB> {
-    pub(crate) store: KVStoreMut<'a, DB>,
+    pub(crate) store: KVStoreMut<'a, DB>, //TODO: does this need to own the store?
```
Comment (NYBACHOK) [inline store/src/types/prefix/mutable.rs:10]: Yes. In next PR you will understand it better. Maybe. Not sure

**Review by NYBACHOK — COMMENTED** [review]
I left a few quick comments, but I need to time to look deeply in code

File: x/ibc/src/ics02_client/types/client_state.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+use crate::types::context::Context;
+
+#[derive(ClientState, Clone, From, TryInto, Debug, Serialize, PartialEq, Deserialize)]
+#[validation(Context<'a, 'b , DB: Database, SK: StoreKey, PSK:ParamsSubspaceKey >)]
+#[execution(Context<'a, 'b, DB: Database, SK:StoreKey, PSK: ParamsSubspaceKey>)]
+// TODO: this enum doesn't serialize to the same JSON as the Cosmos SDK. This is a separate issue to the derive macro issue.
+// E.g. durations are serialized as fields rather than strings, some fields are in camleCase rather than snake_case.
+//#[serde(tag = "@type")] // TODO: uncommenting this causes ClientState derive macro to panic. Uncomment when fixed
+pub enum ClientState {
```
Comment (NYBACHOK) [inline x/ibc/src/ics02_client/types/client_state.rs:24]: `#[serde(untagged)]`?

File: x/ibc/src/ics03_connection/params.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    fn _get_raw_param<DB: Database>(key: &[u8], store: &ImmutablePrefixStore<'_, DB>) -> Vec<u8> {
+        store
+            .get(key)
+            .expect("key should be set in kv store")
+            .clone()
+    }
+
+    fn _parse_param(value: Vec<u8>) -> u64 {
+        String::from_utf8(value)
+            .expect("should be valid utf-8")
```
Comment (NYBACHOK) [inline x/ibc/src/ics03_connection/params.rs:49]: I kinda don't like that. What about `ends_with` and only after that remove?

File: x/ibc/src/types/context/context.rs [hunk 7]
```diff
@@ -0,0 +1,639 @@
...
+pub const KEY_CLIENT_STORE_PREFIX: &str = "clients";
+pub const KEY_CONSENSUS_STATE_PREFIX: &str = "consensusStates";
+pub const KEY_PROCESSED_TIME: &str = "/processedTime";
+
+// KeyProcessedHeight is appended to consensus state key to store the processed height
+const KEY_PROCESSED_HEIGHT: &str = "/processedHeight";
+const KEY_ITERATE_CONSENSUS_STATE_PREFIX: &[u8; 22] = b"iterateConsensusStates";
+
+pub struct Context<'a, 'b, DB, SK, PSK> {
+    pub gears_ctx: &'a mut TxContext<'b, DB, SK>,
```
Comment (NYBACHOK) [inline x/ibc/src/types/context/context.rs:62]: What do you think. Is it better to pass owned version of ctx to xmods or use mutable reference?

**Review by NYBACHOK — APPROVED** [review]

File: docs/tutorials/ibc.md (hunk 3 above)
Comment (joneskm) [reply docs/tutorials/ibc.md:76]: Agreed. The plan is to turn this into documentation at some point. Right now it's more of a cheat sheet.

File: database/src/rocks.rs (hunk 1 above)
Comment (joneskm) [reply database/src/rocks.rs:28]: We should discuss as part of a separate issue. I added it here when I cam across it.

File: Makefile (hunk 2 above)
Comment (joneskm) [reply Makefile:21]: We'll probably move this stuff into Docker containers soon.

File: store/src/types/prefix/mutable.rs (hunk 4 above)
Comment (joneskm) [reply store/src/types/prefix/mutable.rs:10]: :+1:

File: x/ibc/src/ics02_client/types/client_state.rs (hunk 5 above)
Comment (joneskm) [reply x/ibc/src/ics02_client/types/client_state.rs:24]: No, it needs to be tagged to match Cosmos, unfortunately that causes that macro to panic.

File: x/ibc/src/ics03_connection/params.rs (hunk 6 above)
Comment (joneskm) [reply x/ibc/src/ics03_connection/params.rs:49]: We're going to overhaul the handling of params soon. So we can address it then.

File: x/ibc/src/types/context/context.rs (hunk 7 above)
Comment (joneskm) [reply x/ibc/src/types/context/context.rs:62]: For this PR, I believe it has to be a PR. Once your changes have been merged then we could look at changing this.

**Pull request merged**